src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -754,6 +801,31 @@ static MethodSymbol getTopLevelMethod(MethodSymbol method)
             }
         }
 
+        private void AddDeclaredVariable(Symbol symbol, TypeWithAnnotations declaredType) =>
+            AddElementToDictionary(ref _variableTypes, symbol, declaredType, overwriteExisting: true);
+
+        private static void AddElementToDictionary<TKey, TValue>(ref ImmutableDictionary<TKey, TValue> toUpdate, TKey newKey, TValue newValue, bool overwriteExisting = false)
+        {
+            var builder = PooledDictionary<TKey, TValue>.GetInstance();
+            if (toUpdate != null)
+            {
+                foreach (var (key, value) in toUpdate)
+                {
+                    builder.Add(key, value);
+                }
+            }
+
+            if (overwriteExisting)
+            {
+                builder[newKey] = newValue;
+            }
+            else
+            {
+                builder.Add(newKey, newValue);
+            }
+            toUpdate = builder.ToImmutableDictionaryAndFree();
```

Comment:
>toUpdate = builder.ToImmutableDictionaryAndFree(); [](start = 12, length = 50)

This looks unnecessary complex. Could we use ```Set​Item```/```Add``` methods instead?

Reply:
I missed the `SetItem` method.

---
In reply to: [288787926](https://github.com/dotnet/roslyn/pull/35850#discussion_r288787926) [](ancestors = 288787926)

Reply:
Turns out we can't use `SetItem` with TypeSymbols, that ends up calling into its implementation of `Equals(object)`, which throws. I've moved `_variableTypes` back to a `PooledDictionary`

---
In reply to: [288815798](https://github.com/dotnet/roslyn/pull/35850#discussion_r288815798) [](ancestors = 288815798,288787926)